plugins/inputs/t128_graphql/query_builder.go

```diff
@@ -1,20 +1,107 @@
 package t128_graphql
 
-import "strings"
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/Jeffail/gabs"
+)
 
 func buildQuery(entryPoint string, fields map[string]string, tags map[string]string) string {
-	var replacer = strings.NewReplacer("[", "(", "]", "\")", "/", "{", ":", ":\"")
-	query := "query {" + replacer.Replace(entryPoint) + "{"
+	query := "query "
+
+	var buf bytes.Buffer
+	w := io.Writer(&buf)
+	jsonObj := buildQueryObject(entryPoint, fields, tags)
+	traverseQueryObjPreOrder(jsonObj, w)
+
+	query += buf.String()
+	return query
+}
 
-	for _, element := range fields {
-		query += "\n" + element
+func traverseQueryObjPreOrder(jsonObj *gabs.Container, w io.Writer) {
+	jsonChildren, err := jsonObj.ChildrenMap()
+	if err != nil {
+		fmt.Println("error")
+		return
 	}
-	query = strings.TrimSpace(query)
-	for _, element := range tags {
-		query += "\n" + element
+
+	//store and sort the keys
+	keys := make([]string, len(jsonChildren))
+	i := 0
+	for k := range jsonChildren {
+		keys[i] = k
+		i++
 	}
+	sort.Strings(keys)
 
-	query += "\n" + strings.Repeat("}", strings.Count(query, "{"))
+	children := 0
+	for i, key := range keys {
+		if strings.HasPrefix(key, "$") {
+			w.Write([]byte("(" + key[1:] + ":\"" + jsonChildren[key].Data().(string) + "\"){"))
+			continue
+		}
+		children++
+		_, ok := jsonChildren[key].Data().(string)
+		if !ok {
+			if i == 0 {
+				w.Write([]byte("{"))
+			}
+			w.Write([]byte("\n" + key))
+			traverseQueryObjPreOrder(jsonChildren[key], w)
+			continue
+		}
+		if i == 0 {
+			w.Write([]byte("{"))
+		}
+		w.Write([]byte("\n" + key))
+	}
 
-	return query
+	if children > 0 {
+		w.Write([]byte("}"))
+	}
+}
+
+func buildQueryObject(entryPoint string, fields map[string]string, tags map[string]string) *gabs.Container {
+	jsonObj := gabs.New()
+
+	//add items in entry point to obj
+	path := ""
+	pathElements := strings.Split(entryPoint, "/")
+	for _, element := range pathElements {
+		bracketIdx := strings.Index(element, "[")
+		//handle predicates like (name:"ComboEast")
+		if bracketIdx > 0 {
+			path += element[:bracketIdx]
+			predicates := strings.Split(element[bracketIdx+1:len(element)-1], ",")
+			for _, predicate := range predicates {
+				colonIdx := strings.Index(predicate, ":")
+				jsonObj.SetP(predicate[colonIdx+1:], path+".$"+predicate[:colonIdx])
+			}
+			path += "."
+		} else {
+			path += element + "."
+		}
+	}
+
+	//add fields to obj
+	for fieldRenamed, field := range fields {
+		var replacer = strings.NewReplacer("/", ".")
+		partialPath := replacer.Replace(field)
+		jsonObj.SetP(fieldRenamed, path+partialPath)
```

Comment:
This might be a little more performant (not that I'm really worried about performance) and easier to read if you got the "base" json obj above with `path` and then these could be `baseObj.SetP(fieldRenamed, partialPath)`. Not the end of the world though

Reply:
I'll have to revisit this in the next round of refactoring. `SetP` uses a `.` separator while JsonPointer uses a `/` separator so it makes your suggestion non-trivial to implement at the moment